### PR TITLE
Add gaps in time serie

### LINKED ISSUE #2953: Can we add water temperature time series data that has gaps in it?
Some data have gaps in it, for example the attached file:

It would be good to be able to upload these data

Should they be left blank?

image
[WT_E1OLIF-ABDAM_18_11_2008 to 09_02_2011.xlsx](https://github.com/kartoza/django-bims/files/8097137/WT_E1OLIF-ABDAM_18_11_2008.to.09_02_2011.xlsx)

### LINKED ISSUE #2972: Please check - Water temperature data csv download and graphs not matching time series
Please can you check the data uploaded that is only a few months of the year. 

https://freshwaterbiodiversity.org/water-temperature/9980/2011/?taxon=&search=H2ROOI-WORCE&siteId=9980&collector=&category=&yearFrom=&yearTo=&months=&boundary=&userBoundary=&referenceCategory=&spatialFilter=&reference=&endemic=&conservationStatus=%5B%5D&modules=&validated=&sourceCollection=%5B%22fbis%22%5D&abioticData=&ecologicalCategory=&rank=&siteIdOpen=9980&orderBy=name&polygon=&thermalModule=&dst=&startDate=2011-01-01&endDate=2011-12-31

The chart is showing a full year, and then when you download the chart as a csv file, the data is there from 1/1/2011  12:00:00 AM to 6/7/2011  12:00:00 AM, but the data uploaded was only 1/1/2011  12:00:00 AM to 2/2/2011  11:00:00 PM

image

### PR DESCRIPTION
**Pull request by Fanevanjanahary**
fixes #2953
fixes #2972

image

### CODE REVIEW

**Comment by Fanevanjanahary** [comment]
Hi @dimasciput I'm not sure how to calculate the indicators when there is no value, for a moment I just exclude it from the query

File: bims/migrations/0301_auto_20220222_1415.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('bims', '0300_auto_20220127_0253'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='watertemperature',
+            name='value',
+            field=models.FloatField(blank=True, help_text='Water temperature data, if daily data then this is the Mean value.', null=True),
```
Comment (dimasciput) [inline bims/migrations/0301_auto_20220222_1415.py:17]: @Fanevanjanahary please don't make it nullable here.

**Comment by dimasciput** [comment]
@Fanevanjanahary yes please only update it from the frontend, you don't need check if value is null in the backend.

**Review by dimasciput — APPROVED** [review]

File: bims/static/js/non_requirejs/water_temperature.js [hunk 2]
```diff
@@ -67,16 +67,30 @@ function renderWaterTemperatureChart(){
...
+            for (let dataKey in data) {
+                if (dataKey !== 'date_time' && dataKey !== 'days') {
+                    data[dataKey].push([timestamp, null])
+                }
+                data[dataKey].sort((a, b)=> a[0] < b[0] ? -1 : (a[0] > b[0] ? 1 : 0))
+
+            }
+        }
+
+        console.log(data['mean_7'])
```
Comment (dimasciput) [inline bims/static/js/non_requirejs/water_temperature.js:93]: @Fanevanjanahary please remove this

**Review by dimasciput — CHANGES_REQUESTED** [review]

**Comment by dimasciput** [comment]
@Fanevanjanahary can you get all days in the frontend instead to reduce the size of the response from the api?

**Comment by Fanevanjanahary** [comment]
Hi @dimasciput I changed the date range in queryset and now it gets all the data, including the date end

Here is a test from this data [H2ROOI-WORCE 01_12_2008 to 03_02_2011.xlsx](https://github.com/kartoza/django-bims/files/8282469/H2ROOI-WORCE.01_12_2008.to.03_02_2011.xlsx) from @helendallas 

image

image

**Comment by helendallas** [comment]
So, I am not sure we are on the same page here. We are happy to get a shortened time series when data are only for part of a year. We dont need the whole year scale on the x axis. BUT what was happening is the data were only for 2 months, but somehow a graph was generated for the full year.

**Comment by dimasciput** [comment]
@helendallas got it

@Fanevanjanahary let me continue this ticket

**Pull request merged**